File: go/cache/remoteCache/remoteCache.go
// Package remoteCache provides access to a remote cache using
// subset of the memcache protocol.
package remoteCache

import (
	"../../cache"
	"../../protocol"
	"log"
	"net"
)

type remoteCache chan *remoteRequest

type remoteRequest struct {
	cmd                        byte
	opaque                     uint32
	version                    uint64
	header, extras, key, value []byte
	response                   chan *protocol.Packet
}

// New returns a Cache that connects to addr.
// logger, if non-nil, logs the traffic to and from the remote cache.
func New(logger *log.Logger, addr string) (cache.Cache, error) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return nil, err
	}
	c := remoteCache(make(chan *remoteRequest))
	go remoteHandler(logger, c, conn)
	return c, nil
}

func remoteHandler(logger *log.Logger, c remoteCache, conn net.Conn) {
	for {
		req := <-c
		err := protocol.WriteRequest(logger, conn, req.cmd, req.opaque, req.version, req.header, req.extras, req.key, req.value)
		if err != nil {
			req.response <- nil
			continue
		}
		packet, _ := protocol.ReadPacket(logger, conn, protocol.MagicResponse, req.header)
		req.response <- packet
	}
}

func (c remoteCache) Get(key []byte, time cache.Timestamp) ([]byte, uint32, cache.Version, error) {
	req := &remoteRequest{
		cmd:      protocol.CmdGet,
		header:   make([]byte, 24),
		key:      key,
		response: make(chan *protocol.Packet),
	}
	c <- req
	switch packet := <-req.response; {
	case packet == nil:
		break
	case packet.Status() == protocol.StatusNoError:
		return packet.Value, packet.Flags(), cache.Version(packet.Version()), nil
	case packet.Status() == protocol.StatusKeyNotFound:
		return nil, 0, 0, cache.NotFound
	}
	return nil, 0, 0, cache.UnknownError
}

func (c remoteCache) Set(key, data []byte, flags uint32, expires, time cache.Timestamp) (cache.Version, error) {
	req := &remoteRequest{
		cmd:      protocol.CmdSet,
		header:   make([]byte, 24),
		extras:   protocol.MakeFlagsExpiry(flags, uint64(expires), uint64(time)),
		key:      key,
		value:    data,
		response: make(chan *protocol.Packet),
	}
	c <- req
	switch packet := <-req.response; {
	case packet == nil:
		break
	case packet.Status() == protocol.StatusNoError:
		return cache.Version(packet.Version()), nil
	}
	return 0, cache.UnknownError
}

func (c remoteCache) Add(key, data []byte, flags uint32, expires, time cache.Timestamp) (cache.Version, error) {
	req := &remoteRequest{
		cmd:      protocol.CmdAdd,
		header:   make([]byte, 24),
		extras:   protocol.MakeFlagsExpiry(flags, uint64(expires), uint64(time)),
		key:      key,
		value:    data,
		response: make(chan *protocol.Packet),
	}
	c <- req
	switch packet := <-req.response; {
	case packet == nil:
		break
	case packet.Status() == protocol.StatusNoError:
		return cache.Version(packet.Version()), nil
	case packet.Status() == protocol.StatusKeyExists:
		return 0, cache.EntryExists
	}
	return 0, cache.UnknownError
}

func (c remoteCache) Replace(key, data []byte, flags uint32, version cache.Version, expires, time cache.Timestamp) (cache.Version, error) {
	req := &remoteRequest{
		cmd:      protocol.CmdReplace,
		version:  uint64(version),
		header:   make([]byte, 24),
		extras:   protocol.MakeFlagsExpiry(flags, uint64(expires), uint64(time)),
		key:      key,
		value:    data,
		response: make(chan *protocol.Packet),
	}
	c <- req
	switch packet := <-req.response; {
	case packet == nil:
		break
	case packet.Status() == protocol.StatusNoError:
		return cache.Version(packet.Version()), nil
	case packet.Status() == protocol.StatusKeyNotFound:
		return 0, cache.NotFound
	case packet.Status() == protocol.StatusItemNotStored:
		return 0, cache.VersionMismatch
	}
	return 0, cache.UnknownError
}

func (c remoteCache) Delete(key []byte, version cache.Version, time cache.Timestamp) error {
	req := &remoteRequest{
		cmd:      protocol.CmdDelete,
		version:  uint64(version),
		header:   make([]byte, 24),
		key:      key,
		response: make(chan *protocol.Packet),
	}
	c <- req
	switch packet := <-req.response; {
	case packet == nil:
		break
	case packet.Status() == protocol.StatusNoError:
		return nil
	case packet.Status() == protocol.StatusKeyNotFound:
		return cache.NotFound
	case packet.Status() == protocol.StatusItemNotStored:
		return cache.VersionMismatch
	}
	return cache.UnknownError
}

func (c remoteCache) Touch(key []byte, expires cache.Timestamp) error {
	return cache.UnknownError
}

func (c remoteCache) Expire(time cache.Timestamp) {
}

File: hs/RemoteCache.hs
module RemoteCache(
    -- * Types
    RemoteCache,
    -- * Functions
    newRemoteCache
    ) where

import Control.Concurrent(
    Chan,MVar,forkIO,newEmptyMVar,putMVar,takeMVar,newChan,readChan,writeChan)
import Data.ByteString(ByteString)
import qualified Data.ByteString as B
import Data.Time(NominalDiffTime)
import Data.Word(Word32,Word64)
import Network.Socket(
    Socket,SocketType(Stream),HostName,ServiceName,
    connect,socket,getAddrInfo,addrFamily,addrAddress,addrProtocol,defaultProtocol)
import Network.Socket.ByteString(recv,sendMany)

import Cache(
    Cache(get,set,add,replace,delete,expire,touch),
    CacheError(EntryExists,NotFound,VersionMismatch,UnknownError),
    Expires(Expires),Timestamp(Timestamp),Version(Version))
import Logger(Logger)
import Protocol(
    Packet, readPacket, writeRequest,
    packetStatus, packetVersion, packetExtras, packetValue,
    makeExpiry, getFlags, makeFlags,
    magicResponse,
    Cmd, cmdGet, cmdSet, cmdAdd, cmdReplace, cmdDelete,
    statusNoError, statusKeyNotFound, statusKeyExists,
    statusItemNotStored, statusNotSupported, statusInternalError)

-- | RemoteCache accesses a remote cache, using a memcache protocol subset.
data RemoteCache t e = RemoteCache {
    rcSerialize :: e -> (Word32,ByteString),
    rcDeserialize :: (Word32,ByteString) -> e,
    rcConvertExpiry :: t -> t -> NominalDiffTime,
    rcChan :: Chan Req
    }

data Req = Req {
    reqCmd :: Cmd,
    reqVersion :: Word64,
    reqExtras :: [ByteString],
    reqKey :: ByteString,
    reqValue :: Maybe ByteString,
    reqResult :: MVar (Maybe Packet)
    }

-- | Create a new RemoteCache that connects to the given host and port.
-- The Logger logs the requests and responses.  serializeValue and
-- deserializeValue convert the entry values to and from the bytes
-- used with the remote cache.  convertExpiry converts timestamps to
-- NominalDiffTime.
newRemoteCache :: Ord t => Logger -> HostName -> ServiceName -> (e -> (Word32,ByteString)) -> ((Word32,ByteString) -> e) -> (t -> t -> NominalDiffTime) -> IO (RemoteCache t e)
newRemoteCache logger hostname service serializeValue deserializeValue convertExpiry = do
    (addrInfo:_) <- getAddrInfo Nothing (Just hostname) (Just service)
    sock <- socket (addrFamily addrInfo) Stream defaultProtocol
    connect sock (addrAddress addrInfo)
    chan <- newChan
    forkIO (process sock chan)
    return RemoteCache {
        rcSerialize = serializeValue,
        rcDeserialize = deserializeValue,
        rcConvertExpiry = convertExpiry,
        rcChan = chan
        }
  where
    process sock chan = do
        req <- readChan chan
        writeRequest logger sock (reqCmd req) 0 (reqVersion req) (reqExtras req) (Just $ reqKey req) (reqValue req)
        packet <- readPacket logger sock magicResponse
        putMVar (reqResult req) packet
        process sock chan

instance Cache RemoteCache where
    get RemoteCache { rcDeserialize = deser, rcChan = chan } key _ = do
        mVar <- newEmptyMVar
        writeChan chan Req {
                        reqCmd = cmdGet,
                        reqVersion = 0,
                        reqExtras = [],
                        reqKey = key,
                        reqValue = Nothing,
                        reqResult = mVar
                        }
        maybePacket <- takeMVar mVar
        return $ makeResult (fmap packetStatus maybePacket) (fmap packetVersion maybePacket) (fmap packetExtras maybePacket) (fmap packetValue maybePacket)
      where
        makeResult (Just status) (Just version) (Just extras) (Just (Just value))
          | status == statusNoError =
                Right (deser (getFlags (B.concat extras) 0, value), Version version)
          | status == statusKeyNotFound = Left NotFound
          | otherwise = Left UnknownError
        makeResult (Just status) _ _ _
          | status == statusKeyNotFound = Left NotFound
          | otherwise = Left UnknownError
        makeResult _ _ _ _ = Left UnknownError

    set RemoteCache { rcSerialize = ser, rcConvertExpiry = expiry, rcChan = chan } key e (Expires exp) (Timestamp t) = do
        let (flags,value) = ser e
        mVar <- newEmptyMVar
        writeChan chan Req {
                        reqCmd = cmdSet,
                        reqVersion = 0,
                        reqExtras = [makeFlags flags, makeExpiry $ expiry exp t],
                        reqKey = key,
                        reqValue = Just value,
                        reqResult = mVar
                        }
        maybePacket <- takeMVar mVar
        return $ makeResult (fmap packetStatus maybePacket) (fmap packetVersion maybePacket)
      where
        makeResult (Just status) (Just version)
          | status == statusNoError = Right (Version version)
          | otherwise = Left UnknownError
        makeResult _ _ = Left UnknownError

    add RemoteCache { rcSerialize = ser, rcConvertExpiry = expiry, rcChan = chan } key e (Expires exp) (Timestamp t) = do
        let (flags,value) = ser e
        mVar <- newEmptyMVar
        writeChan chan Req {
                        reqCmd = cmdAdd,
                        reqVersion = 0,
                        reqExtras = [makeFlags flags, makeExpiry $ expiry exp t],
                        reqKey = key,
                        reqValue = Just value,
                        reqResult = mVar
                        }
        maybePacket <- takeMVar mVar
        return $ makeResult (fmap packetStatus maybePacket) (fmap packetVersion maybePacket)
      where
        makeResult (Just status) (Just version)
          | status == statusNoError = Right (Version version)
          | status == statusKeyExists = Left EntryExists
          | otherwise = Left UnknownError
        makeResult _ _ = Left UnknownError

    replace RemoteCache { rcSerialize = ser, rcConvertExpiry = expiry, rcChan = chan } key (Version version) e (Expires exp) (Timestamp t) = do
        let (flags,value) = ser e
        mVar <- newEmptyMVar
        writeChan chan Req {
                        reqCmd = cmdReplace,
                        reqVersion = version,
                        reqExtras = [makeFlags flags, makeExpiry $ expiry exp t],
                        reqKey = key,
                        reqValue = Just value,
                        reqResult = mVar
                        }
        maybePacket <- takeMVar mVar
        return $ makeResult (fmap packetStatus maybePacket) (fmap packetVersion maybePacket)
      where
        makeResult (Just status) (Just version)
          | status == statusNoError = Right (Version version)
          | status == statusKeyNotFound = Left NotFound
          | status == statusItemNotStored = Left VersionMismatch
          | otherwise = Left UnknownError
        makeResult _ _ = Left UnknownError

    delete RemoteCache { rcChan = chan } key (Version version) _ = do
        mVar <- newEmptyMVar
        writeChan chan Req {
                        reqCmd = cmdDelete,
                        reqVersion = version,
                        reqExtras = [],
                        reqKey = key,
                        reqValue = Nothing,
                        reqResult = mVar
                        }
        maybePacket <- takeMVar mVar
        return $ makeResult (fmap packetStatus maybePacket)
      where
        makeResult (Just status)
          | status == statusNoError = Nothing
          | status == statusKeyNotFound = Just NotFound
          | status == statusItemNotStored = Just VersionMismatch
          | otherwise = Just UnknownError
        makeResult _ = Just UnknownError

    touch _ _ _ = return (Left UnknownError)
    expire _ _ = return ()
